### ISSUE TITLE
Runtime error with mutually recursive records

### ISSUE THREAD

**Issue by TheSeamau5** [elm-lang/elm-compiler#912]
The following code

```elm
p = { x = q.x, y = 2 }
q = { x = 3, y = p.y }
```

yields:

```
Cannot read property 'x' of undefined


Open the developer console for more details.
```

In the example, `p` and `q` mutually depend on each other but in some sense their values can be "resolved" (at least in the logic programming sense).

Allowing this behavior may cause headaches (as it may involve initialization for all types and special getters or whatever).

At the very least this behavior should trigger a compiler warning that looks something like:

```
Line 3: p and q are defined in terms of each other. 
Instead, you may want to unwind the recursion as follows:

r = { x = 3, y = 2 }

p = { x = r.x, y = r.y }
q = { x = r.x, y = r.y }
```

This may sound hard to do, so perhaps in the meantime, we could return an error like:

```
Line 3: p defined in terms of q and q defined in terms of p. Please fix recursion.
```

or something nicer.

**Comment by evancz** [comment]
This appears to be a duplicate of #873, so lets track it from there! Seems like this may get resolved with the work on dead code elimination. We shall see!

**Comment by srikumarks** [comment]
Can we arrange for this kind of initialization to be compiled into something like this?

```js
p = {}
q = {}
p.y = 2;
q.x = 3;
p.x = q.x;
q.y = p.y;
```
